File: ios/FindGlasses/Finder/YoloDetector.swift
import CoreGraphics
import Foundation
import onnxruntime_objc

/// ONNX Runtime 으로 도는 YOLO-World s. Android `YoloDetector.kt` 와 같은 후처리.
///
/// 어휘가 검출 헤드에 구워져 있어 런타임에 텍스트 인코더가 없다.
/// 출력은 [1, 4+nc, N]
///   채널 0..3     박스 xywh — 입력 픽셀 기준, 정규화 아님
///   채널 4..4+nc  클래스 점수 (sigmoid 적용됨)
///
/// 라이선스: 모델 가중치는 Ultralytics AGPL-3.0. 앱 소스 공개 전제.
final class YoloDetector {

    let provider: String
    let inputSize: Int

    /// 찾을 대상의 클래스 번호. 나머지 어휘는 네거티브가 된다.
    var targetClasses: Set<Int> = ModelCatalog.Target.glasses.classes

    private let env: ORTEnv
    private let session: ORTSession
    private let inputName: String
    private let outputName: String
    private let preprocess: Preprocess
    private let nc = ModelCatalog.classes.count

    private static let maxDetections = 5
    /// 후보가 이미 남긴 박스 안에 이 비율 이상 들어가면 같은 물체로 본다
    private static let containment: CGFloat = 0.65

    init(modelPath: String, inputSize: Int) throws {
        self.inputSize = inputSize
        env = try ORTEnv(loggingLevel: .warning)

        // Core ML 을 먼저 시도하고 안 되면 CPU. 아이폰은 Neural Engine 이 있어
        // Android(NNAPI ≈ CPU) 와 달리 차이가 클 수 있다 — 실기기에서 잰다.
        var created: (ORTSession, String)?
        do {
            let opts = try ORTSessionOptions()
            try opts.appendCoreMLExecutionProvider(with: ORTCoreMLExecutionProviderOptions())
            created = (try ORTSession(env: env, modelPath: modelPath, sessionOptions: opts), "CoreML")
        } catch {
            NSLog("[YoloDetector] CoreML 실패, CPU 로 전환: \(error)")
        }
        if created == nil {
            let opts = try ORTSessionOptions()
            try opts.setIntraOpNumThreads(4)
            created = (try ORTSession(env: env, modelPath: modelPath, sessionOptions: opts), "CPU")
        }
        session = created!.0
        provider = created!.1

        inputName = try session.inputNames().first ?? "images"
        outputName = try session.outputNames().first ?? "output0"
        guard let p = Preprocess(size: inputSize) else {
            throw NSError(domain: "YoloDetector", code: 1, userInfo: [NSLocalizedDescriptionKey: "전처리 버퍼 생성 실패"])
        }
        preprocess = p
    }

    func detect(_ image: CGImage, threshold: Float) throws -> DetectResult {
        let t0 = CFAbsoluteTimeGetCurrent()
        let fit = preprocess.run(image)
        let t1 = CFAbsoluteTimeGetCurrent()

        let s = NSNumber(value: inputSize)
        let data = preprocess.chw.withUnsafeBufferPointer { NSMutableData(bytes: $0.baseAddress, length: $0.count * MemoryLayout<Float>.size) }
        let input = try ORTValue(tensorData: data, elementType: .float, shape: [1, 3, s, s])
        let outputs = try session.run(withInputs: [inputName: input], outputNames: [outputName], runOptions: nil)
        let t2 = CFAbsoluteTimeGetCurrent()

        guard let out = outputs[outputName] else {
            throw NSError(domain: "YoloDetector", code: 2, userInfo: [NSLocalizedDescriptionKey: "출력 없음"])
        }
        let shape = try out.tensorTypeAndShapeInfo().shape.map { $0.intValue }
        let tensor = try out.tensorData() as Data
        let (dets, top) = parse(tensor, anchors: shape[2], fit: fit, threshold: threshold)
        let t3 = CFAbsoluteTimeGetCurrent()

        return DetectResult(
            detections: dets,
            inferenceMs: (t2 - t1) * 1000,
            preprocessMs: (t1 - t0) * 1000,
            postprocessMs: (t3 - t2) * 1000,
            topPositiveScore: top
        )
    }

    private func parse(_ data: Data, anchors: Int, fit: Preprocess.Fit, threshold: Float) -> ([Detection], Float) {
        let targets = targetClasses
        var raw: [Detection] = []
        var topPositive: Float = 0

        data.withUnsafeBytes { (ptr: UnsafeRawBufferPointer) in
            let buf = ptr.bindMemory(to: Float.self)
            for i in 0..<anchors {
                var bestClass = -1
                var bestScore: Float = 0
                var positive: Float = 0
                for c in 0..<nc {
                    let v = buf[(4 + c) * anchors + i]
                    if targets.contains(c), v > positive { positive = v }
                    if v > bestScore { bestScore = v; bestClass = c }
                }
                if positive > topPositive { topPositive = positive }

                // 다른 어휘가 더 높으면 대상이 아니다
                guard targets.contains(bestClass), bestScore >= threshold else { continue }

                let box = toSourceRect(
                    cx: CGFloat(buf[i]), cy: CGFloat(buf[anchors + i]),
                    w: CGFloat(buf[2 * anchors + i]), h: CGFloat(buf[3 * anchors + i]), fit: fit)
                raw.append(Detection(score: bestScore, box: box, classIndex: bestClass))
            }
        }
        raw.sort { $0.score > $1.score }
        return (nms(raw), topPositive)
    }

    /// letterbox 입력 픽셀 xywh → 똑바로 세운 원본 프레임 픽셀
    private func toSourceRect(cx: CGFloat, cy: CGFloat, w: CGFloat, h: CGFloat, fit: Preprocess.Fit) -> CGRect {
        let px = CGFloat(fit.padX), py = CGFloat(fit.padY)
        let x1 = ((cx - w / 2 - px) / fit.scale).clamped(0, CGFloat(fit.srcW))
        let x2 = ((cx + w / 2 - px) / fit.scale).clamped(0, CGFloat(fit.srcW))
        let y1 = ((cy - h / 2 - py) / fit.scale).clamped(0, CGFloat(fit.srcH))
        let y2 = ((cy + h / 2 - py) / fit.scale).clamped(0, CGFloat(fit.srcH))
        return CGRect(x: x1, y: y1, width: x2 - x1, height: y2 - y1)
    }

    /// 같은 안경에 앵커 여러 개가 겹쳐 붙는다. IoU 와 포함 관계로 누른다.
    private func nms(_ sorted: [Detection], iouThreshold: CGFloat = 0.3) -> [Detection] {
        var kept: [Detection] = []
        for d in sorted {
            if !kept.contains(where: { suppresses($0.box, d.box, iouThreshold) }) { kept.append(d) }
            if kept.count >= Self.maxDetections { break }
        }
        return kept
    }

    private func suppresses(_ kept: CGRect, _ cand: CGRect, _ iouThreshold: CGFloat) -> Bool {
        if iou(kept, cand) >= iouThreshold { return true }
        let inter = kept.intersection(cand)
        let candArea = cand.width * cand.height
        if candArea <= 0 { return true }
        if inter.isNull { return false }
        return inter.width * inter.height / candArea >= Self.containment
    }
}

func iou(_ a: CGRect, _ b: CGRect) -> CGFloat {
    let inter = a.intersection(b)
    if inter.isNull || inter.width <= 0 || inter.height <= 0 { return 0 }
    let i = inter.width * inter.height
    let u = a.width * a.height + b.width * b.height - i
    return u > 0 ? i / u : 0
}

extension Comparable {
    func clamped(_ lo: Self, _ hi: Self) -> Self { min(max(self, lo), hi) }
}

File: android/app/src/main/java/com/myaune/findglasses/YoloDetector.kt
package com.myaune.findglasses

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import android.graphics.Bitmap
import android.graphics.RectF
import android.util.Log
import java.io.File

/**
 * ONNX Runtime 으로 도는 YOLO 계열 (YOLO-World / YOLOE).
 *
 * 어휘가 검출 헤드에 구워져 있어 런타임에 텍스트 인코더가 없다.
 * dynamic 으로 익스포트했으므로 입력 해상도를 생성 시점에 고른다.
 *
 * 출력은 [1, C, N]
 *   채널 0..3        박스 xywh — 입력 픽셀 기준, 정규화 아님
 *   채널 4..4+nc     클래스 점수 (sigmoid 적용됨)
 *   그 뒤            YOLOE(seg) 는 마스크 계수 32 채널이 더 붙는다
 *
 * nc 를 C-4 로 역산하면 YOLOE 에서 39개로 잘못 읽는다. CLASSES.size 를 쓴다.
 *
 * 라이선스: 모델 GPL-3.0, ultralytics AGPL-3.0. 앱 소스 공개 전제.
 */
class YoloDetector private constructor(
    private val env: OrtEnvironment,
    private val session: OrtSession,
    override val provider: String,
    override val modelBytes: Long,
    override val label: String,
    private val inputSize: Int,
) : Detector {

    private val preprocess = Preprocess(
        size = inputSize,
        padGray = ModelCatalog.PAD_GRAY,
        // YOLO 는 /255 만 한다. ImageNet 정규화를 넣으면 검출이 무너진다.
        mean = null,
        std = null,
    )
    private val inputName: String = session.inputNames.first()
    private val nc = ModelCatalog.CLASSES.size

    @Volatile
    override var targetClasses: IntArray = ModelCatalog.Target.GLASSES.classes

    override fun close() = session.close()

    override fun detect(src: Bitmap, rotationDegrees: Int, threshold: Float): DetectResult {
        val t0 = System.nanoTime()
        val (buffer, fit) = preprocess.run(src, rotationDegrees)
        val t1 = System.nanoTime()

        val s = inputSize.toLong()
        val tensor = OnnxTensor.createTensor(env, buffer, longArrayOf(1, 3, s, s))
        val out = tensor.use { session.run(mapOf(inputName to it)) }
        val t2 = System.nanoTime()

        val parsed = out.use { parse(it, fit, threshold) }
        val t3 = System.nanoTime()

        return DetectResult(
            detections = parsed.first,
            inferenceMs = (t2 - t1) / 1_000_000,
            preprocessMs = (t1 - t0) / 1_000_000,
            postprocessMs = (t3 - t2) / 1_000_000,
            topPositiveScore = parsed.second,
        )
    }

    private fun parse(
        out: OrtSession.Result,
        fit: Preprocess.Fit,
        threshold: Float,
    ): Pair<List<Detection>, Float> {
        val t = out[0] as OnnxTensor
        val anchors = t.info.shape[2].toInt()
        val buf = t.floatBuffer

        val raw = ArrayList<Detection>()
        var topPositive = 0f
        val targets = targetClasses

        for (i in 0 until anchors) {
            var bestClass = -1
            var bestScore = 0f
            var positiveScore = 0f

            for (c in 0 until nc) {
                val v = buf.get((4 + c) * anchors + i)
                if (c in targets && v > positiveScore) positiveScore = v
                if (v > bestScore) {
                    bestScore = v
                    bestClass = c
                }
            }
            if (positiveScore > topPositive) topPositive = positiveScore

            // 다른 어휘가 더 높으면 대상이 아니다. 안경을 찾을 때는 열쇠·컵 등이,
            // 열쇠를 찾을 때는 안경이 네거티브가 된다.
            if (bestClass !in targets || bestScore < threshold) continue

            raw.add(
                Detection(
                    bestScore,
                    toSourceRect(
                        buf.get(0 * anchors + i), buf.get(1 * anchors + i),
                        buf.get(2 * anchors + i), buf.get(3 * anchors + i), fit,
                    ),
                    bestClass,
                )
            )
        }

        raw.sortByDescending { it.score }
        return nms(raw) to topPositive
    }

    /** letterbox 입력 픽셀 xywh → 회전 보정된 원본 프레임 픽셀 */
    private fun toSourceRect(
        cx: Float, cy: Float, w: Float, h: Float, fit: Preprocess.Fit,
    ): RectF {
        var x1 = (cx - w / 2 - fit.padX) / fit.scale
        var y1 = (cy - h / 2 - fit.padY) / fit.scale
        var x2 = (cx + w / 2 - fit.padX) / fit.scale
        var y2 = (cy + h / 2 - fit.padY) / fit.scale
        x1 = x1.coerceIn(0f, fit.srcW.toFloat())
        x2 = x2.coerceIn(0f, fit.srcW.toFloat())
        y1 = y1.coerceIn(0f, fit.srcH.toFloat())
        y2 = y2.coerceIn(0f, fit.srcH.toFloat())
        return RectF(x1, y1, x2, y2)
    }

    /**
     * 같은 안경에 앵커 여러 개가 겹쳐 붙는다. 누르지 않으면 Tracker 가 이를
     * "여러 번 검출" 로 받아 단일 프레임만으로 발화한다. 기획서 9절의 3회는
     * 3개 프레임을 뜻한다.
     */
    private fun nms(sorted: List<Detection>, iouThreshold: Float = 0.3f): List<Detection> {
        val kept = ArrayList<Detection>()
        for (d in sorted) {
            if (kept.none { suppresses(it.box, d.box, iouThreshold) }) kept.add(d)
            if (kept.size >= MAX_DETECTIONS) break
        }
        return kept
    }

    /**
     * 이미 남긴 박스가 새 박스를 덮어쓰는가.
     *
     * IoU 만 보면 한 안경 위에 박스가 여러 개 남는다. 한쪽 렌즈만 잡은 것,
     * 양쪽을 잡은 것, 다리까지 포함한 것의 IoU 가 0.5 미만이라 다 살아남았다.
     * 그래서 포함 관계도 본다 — 작은 박스가 큰 박스 안에 대부분 들어가 있으면
     * 같은 물체로 보고 버린다.
     */
    private fun suppresses(kept: RectF, cand: RectF, iouThreshold: Float): Boolean {
        if (iou(kept, cand) >= iouThreshold) return true
        val ix = maxOf(0f, minOf(kept.right, cand.right) - maxOf(kept.left, cand.left))
        val iy = maxOf(0f, minOf(kept.bottom, cand.bottom) - maxOf(kept.top, cand.top))
        val inter = ix * iy
        val candArea = cand.width() * cand.height()
        if (candArea <= 0f) return true
        return inter / candArea >= CONTAINMENT
    }

    private fun iou(a: RectF, b: RectF): Float {
        val ix = maxOf(0f, minOf(a.right, b.right) - maxOf(a.left, b.left))
        val iy = maxOf(0f, minOf(a.bottom, b.bottom) - maxOf(a.top, b.top))
        val inter = ix * iy
        if (inter <= 0f) return 0f
        val union = a.width() * a.height() + b.width() * b.height() - inter
        return if (union > 0f) inter / union else 0f
    }

    companion object {
        private const val TAG = "YoloDetector"
        private const val MAX_DETECTIONS = 5

        /** 후보가 이미 남긴 박스 안에 이 비율 이상 들어가면 같은 물체로 본다. */
        private const val CONTAINMENT = 0.65f

        fun create(
            modelFile: File,
            label: String,
            inputSize: Int,
            preferNnapi: Boolean,
        ): YoloDetector {
            require(modelFile.exists()) { "모델 파일이 없습니다: ${modelFile.absolutePath}" }
            val env = OrtEnvironment.getEnvironment()

            if (preferNnapi) {
                try {
                    val opts = OrtSession.SessionOptions()
                    opts.addNnapi()
                    val s = env.createSession(modelFile.absolutePath, opts)
                    Log.i(TAG, "$label NNAPI")
                    return YoloDetector(env, s, "NNAPI", modelFile.length(), label, inputSize)
                } catch (e: Throwable) {
                    Log.w(TAG, "NNAPI 실패, CPU 로 전환: ${e.message}")
                }
            }

            val opts = OrtSession.SessionOptions()
            opts.setIntraOpNumThreads(4)   // Cortex-A73 ×4
            val s = env.createSession(modelFile.absolutePath, opts)
            Log.i(TAG, "$label CPU")
            return YoloDetector(env, s, "CPU", modelFile.length(), label, inputSize)
        }
    }
}
